arangod/Aql/ExecutionBlockImpl.cpp

```diff
@@ -1820,6 +1855,82 @@ auto ExecutionBlockImpl<Executor>::testInjectInputRange(DataRange range, SkipRes
 }
 #endif
 
+template <class Executor>
+ExecutionBlockImpl<Executor>::Context::Context(ExecutionBlockImpl& block, AqlCallStack const& callstack)
+    : stack(callstack), clientCallList(this->stack.popCall()) {
+  if constexpr (std::is_same_v<Executor, SubqueryEndExecutor>) {
+    // In subqeryEndExecutor we actually manage two calls.
+    // The clientCall defines what will go into the Executor.
+    // on SubqueryEnd this call is generated based on the call from downstream
+
+    if (block._outputItemRow != nullptr && block._outputItemRow->isInitialized()) {
+      // If we return with a waiting state, we need to report it to the
+      // subquery callList, but not pull it of.
+      auto& subQueryCall = clientCallList.modifyNextCall();
+      // Overwrite with old state.
+      subQueryCall = block._outputItemRow->getClientCall();
+    }
+    stack.pushCall(std::move(clientCallList));
+    clientCallList = AqlCallList{AqlCall{}, AqlCall{}};
+    clientCall = clientCallList.popNextCall();
+  } else {
+    clientCall = clientCallList.popNextCall();
+  }
+  // We got called with a skip count already set!
+  // Caller is wrong fix it.
+  TRI_ASSERT(clientCall.getSkipCount() == 0);
+  
+  TRI_ASSERT(!(clientCall.getOffset() == 0 && clientCall.softLimit == AqlCall::Limit{0u}));
+  TRI_ASSERT(!(clientCall.hasSoftLimit() && clientCall.fullCount));
+  TRI_ASSERT(!(clientCall.hasSoftLimit() && clientCall.hasHardLimit()));
+}
+
+template <class Executor>
+bool ExecutionBlockImpl<Executor>::PrefetchTask::isConsumed() const {
+  return state.load(std::memory_order_relaxed) == State::Consumed;
+}
+
+template <class Executor>
+bool ExecutionBlockImpl<Executor>::PrefetchTask::tryClaim() {
+  auto expected = State::Pending;
+  return state.load(std::memory_order_relaxed) == expected &&
+    state.compare_exchange_strong(expected, State::InProgress, std::memory_order_relaxed);
+}
+
+template <class Executor>
+void ExecutionBlockImpl<Executor>::PrefetchTask::waitFor() {
```

Comment:
This method is called in destructors, which requires noexcept-ness. 
I think the method is effectively `noexcept`, although `guard` and `bell` may throw under some adverse conditions. 
We could either wrap this method in a try...catch, or mark it `noexcept` and have `std::terminate()` kill the process if an exception escapes from here. What shall we do, if anything?

Reply:
Since `std::mutex` and `std::condition_variable` can potentially throw, I would not mark this function as `noexcept`, but add a try-catch in the destructor. IMHO that would be the cleanest solution.

Reply:
Fine for me.

Reply:
Actually, after checking the standard under which circumstances `std::mutex` might throw, I would go with your last suggestion and mark the function `noexcept` and have the process terminate if we really have an exception.